### PR TITLE
acado/1.2.2-beta template fix

### PR DESCRIPTION
**Pull request by blackliner**
Specify library name and version:  **acado/1.2.2-beta**

The installed file `include/acado/code_generation/templates/templates.hpp` contains hardcoded paths that need to be adapted. Otherwise we get CCI CI paths like `/home/conan/w/cci_PR-3967/.conan/data/acado/1.2.2-beta/_/_/build/6384f4f1a20d5949e786ef62bf18a6d2899555e4/source_subfolder/acado/code_generation/export_templated_file.cpp` 😕 

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by madebr** [comment]
Would it be possible to apply a patch before the build that changes `TEMPLATE_PATHS`?
https://github.com/acado/acado/blob/b4e28f3131f79cadfd1a001e9fff061f361d3a0f/acado/code_generation/templates/templates.hpp.in#L4

**Comment by blackliner** [comment]
> Would it be possible to apply a patch before the build that changes `TEMPLATE_PATHS`?
> https://github.com/acado/acado/blob/b4e28f3131f79cadfd1a001e9fff061f361d3a0f/acado/code_generation/templates/templates.hpp.in#L4

No, unfortunately. The path has to point to the folder where the package is installed, which is only known when `conan install ..` is called.

**Comment by madebr** [comment]
Looking at the cmake script, they add both the current source directory and the install directory:
https://github.com/acado/acado/blob/b4e28f3131f79cadfd1a001e9fff061f361d3a0f/acado/code_generation/templates/CMakeLists.txt#L76-L80

This function will also not work anymore when moving the built libraries between machines with different folder structure:
https://github.com/acado/acado/blob/b4e28f3131f79cadfd1a001e9fff061f361d3a0f/acado/utils/acado_io_utils.cpp#L228
(because the paths are hard-coded built-in)

This looks hacky/bad design.
Maybe they are open to better suggestions? (probably not as the project is non-active)

How about the following?
Wrap all access to `TEMPLATE_PATHS` in a function, named e.g. `acode_template_paths()` which does the following:
```c++
std::string
acado_template_paths() {
    if (::getenv("ACADO_TEMPLATE_PATHS")) {
        return ::getenv("ACADO_TEMPLATE_PATHS");
    }
    return TEMPLATE_PATHS
}
```
And add `self.env_info.ACADO_TEMPLATE_PATHS = os.path.join(self.package_folder, "include", "acado", "code_generation", "templates")` to `package_info` of acado?

**Comment by blackliner** [comment]
> Looking at the cmake script, they add both the current source directory and the install directory:
> https://github.com/acado/acado/blob/b4e28f3131f79cadfd1a001e9fff061f361d3a0f/acado/code_generation/templates/CMakeLists.txt#L76-L80

Yes, not necessary in my opinion 🤷 

> This function will also not work anymore when moving the built libraries between machines with different folder structure:
> https://github.com/acado/acado/blob/b4e28f3131f79cadfd1a001e9fff061f361d3a0f/acado/utils/acado_io_utils.cpp#L228
> (because the paths are hard-coded built-in)

Yes, bad design IMHO

> This looks hacky/bad design.
> Maybe they are open to better suggestions? (probably not as the project is non-active)

Totally agree, very bad for packaging.

> How about the following?
> Wrap all access to `TEMPLATE_PATHS` in a function, named e.g. `acode_template_paths()` which does the following:
> 
> ```c++
> std::string
> acado_template_paths() {
>     if (::getenv("ACADO_TEMPLATE_PATHS")) {
>         return ::getenv("ACADO_TEMPLATE_PATHS");
>     }
>     return TEMPLATE_PATHS
> }
> ```
> 
> And add `self.env_info.ACADO_TEMPLATE_PATHS = os.path.join(self.package_folder, "include", "acado", "code_generation", "templates")` to `package_info` of acado?

I do not know how `self.env_info` works. How should this env var be made available for a consumer that is using `cmake_find_package*` generators? In my opinion the real fix would be to move the [`configure_file`](https://github.com/acado/acado/blob/v1.2.2beta/acado/code_generation/templates/CMakeLists.txt#L90) from install to configure/build time. I think this could be technically possible, because all those files are not needed during the initial compilation of the library itself, but only for generating the code generator executable.

This change would then be even more hacky, but it might be possible to relocate the install if done properly 🤔

**Comment by madebr** [comment]
> I do not know how `self.env_info` works. How should this env var be made available for a consumer that is using `cmake_find_package*` generators?

It is a runtime variable.
It is set as:
```python
def package_info(self):
    self.env_info.OCADA_TEMPLATE_PATHS = os.path.join(self.package_folder, "include", "acado", "code_generation", "templates")
```
What it does is setting the `OCADA_TEMPLATE_PATHS` environment variable when used inside conan or when using the [`virtualenv`](https://docs.conan.io/en/latest/reference/generators/virtualenv.html) generator.

> In my opinion the real fix would be to move the [`configure_file`](https://github.com/acado/acado/blob/v1.2.2beta/acado/code_generation/templates/CMakeLists.txt#L90) from install to configure/build time. I think this could be technically possible, because all those files are not needed during the initial compilation of the library itself, but only for generating the code generator executable.

What should be avoided is that the paths string is embedded in the library.
So when you run `strings` on the shared library, no string containing the template path should be found.
```
# This should have no matches (or unrelated ones)
strings libacado.so | grep code_generation`
```

> This change would then be even more hacky, but it might be possible to relocate the install if done properly 🤔
> 
> EDIT: maybe it is possible to run just this one [CMakeLists.txt](https://github.com/acado/acado/blob/v1.2.2beta/acado/code_generation/templates/CMakeLists.txt) again during the `find_package( acado )` call?

How about this?
```c++
static std::string acado_template_paths = TEMPLATE_PATHS;
std::string
acado_get_template_paths() {
    return acado_template_paths 
}
std::string
set_acado_template_paths(const std::string &paths) {
    acado_template_paths = paths;
}
```
This keeps the embedded paths, but lets the user override the template paths (which should be done only once at initialization).

**Comment by blackliner** [comment]
> > I do not know how `self.env_info` works. How should this env var be made available for a consumer that is using `cmake_find_package*` generators?
> 
> It is a runtime variable.
> It is set as:
> 
> ```python
> def package_info(self):
>     self.env_info.OCADA_TEMPLATE_PATHS = os.path.join(self.package_folder, "include", "acado", "code_generation", "templates")
> ```
> 
> What it does is setting the `OCADA_TEMPLATE_PATHS` environment variable when used inside conan or when using the [`virtualenv`](https://docs.conan.io/en/latest/reference/generators/virtualenv.html) generator.

So this will not work with "native" CMake usage, aka no conan related code inside the consumers `CMakeLists.txt` ?

**Comment by madebr** [comment]
> > What it does is setting the `OCADA_TEMPLATE_PATHS` environment variable when used inside conan or when using the [`virtualenv`](https://docs.conan.io/en/latest/reference/generators/virtualenv.html) generator.
> 
> So this will not work with "native" CMake usage, aka no conan related code inside the consumers `CMakeLists.txt` ?

My `self.env_info.OCADA_TEMPLATE_PATHS` don't need anything when building, only when running ocada.
If ocada is run inside CMake, then `$ENV{OCADA_TEMPLATE_PATHS}` must be set.


> EDIT: I also tried out running the `configure_file` again during `find_package`, but this will be even hackier and did not work after 2x more changed code.

Not a real fan of this.
That's why I suggest adding the get/setter functions for the template paths.

> EDIT2: Also keep in mind, `TEMPLATE_PATHS` is only used for code that is part of the code generator executable. It is not part of the library itself! See the only other usage of it in [acadoCopyTemplateFile](https://github.com/acado/acado/search?q=acadoCopyTemplateFile), which is itself not used anymore.

[This file in acado/utils uses the TEMPLATE_PATHS definition](https://github.com/acado/acado/blob/b4e28f3131f79cadfd1a001e9fff061f361d3a0f/acado/utils/acado_io_utils.cpp#L228).
The [`utils` directory](https://github.com/acado/acado/blob/b4e28f3131f79cadfd1a001e9fff061f361d3a0f/acado/CMakeLists.txt#L33) is [globbed](https://github.com/acado/acado/blob/b4e28f3131f79cadfd1a001e9fff061f361d3a0f/acado/CMakeLists.txt#L109-L112) and added to the [`acado_toolkit` library](https://github.com/acado/acado/blob/b4e28f3131f79cadfd1a001e9fff061f361d3a0f/acado/CMakeLists.txt#L115).

**Comment by blackliner** [comment]
> My `self.env_info.OCADA_TEMPLATE_PATHS` don't need anything when building, only when running ocada.
> If ocada is run inside CMake, then `$ENV{OCADA_TEMPLATE_PATHS}` must be set.

And this is what I want to avoid, having to alter the way this library is used, forcing consumers to provide arbitrary paths.

> > EDIT2: Also keep in mind, `TEMPLATE_PATHS` is only used for code that is part of the code generator executable. It is not part of the library itself! See the only other usage of it in [acadoCopyTemplateFile](https://github.com/acado/acado/search?q=acadoCopyTemplateFile), which is itself not used anymore.
> 
> [This file in acado/utils uses the TEMPLATE_PATHS definition](https://github.com/acado/acado/blob/b4e28f3131f79cadfd1a001e9fff061f361d3a0f/acado/utils/acado_io_utils.cpp#L228).
> The [`utils` directory](https://github.com/acado/acado/blob/b4e28f3131f79cadfd1a001e9fff061f361d3a0f/acado/CMakeLists.txt#L33) is [globbed](https://github.com/acado/acado/blob/b4e28f3131f79cadfd1a001e9fff061f361d3a0f/acado/CMakeLists.txt#L109-L112) and added to the [`acado_toolkit` library](https://github.com/acado/acado/blob/b4e28f3131f79cadfd1a001e9fff061f361d3a0f/acado/CMakeLists.txt#L115).

Please have a look at the link I posted earlier: https://github.com/acado/acado/search?q=acadoCopyTemplateFile
The function `acadoCopyTemplateFile` is NOT used anymore.

**Comment by madebr** [comment]
> And this is what I want to avoid, having to alter the way this library is used, forcing consumers to provide arbitrary paths.

It's not arbitrary.
In cmake it can be set as `acada_ROOT/include/templates` (or something similar).
But it's weird nonetheless.
Anyhow, when I want to inform my users about some path, I would pass it through a cmake variable, or section in a `.pc` file.

> Please have a look at the link I posted earlier: https://github.com/acado/acado/search?q=acadoCopyTemplateFile
> The function `acadoCopyTemplateFile` is NOT used anymore.

You're right about `acadoCOpyTemplateFile`.
The only other user `cpp` user of `TEMPLATE_PATHS` is this one: https://github.com/acado/acado/blob/b4e28f3131f79cadfd1a001e9fff061f361d3a0f/acado/code_generation/export_templated_file.cpp#L69
Is that method called somewhere in the library?

**Comment by blackliner** [comment]
> > And this is what I want to avoid, having to alter the way this library is used, forcing consumers to provide arbitrary paths.
> 
> It's not arbitrary.
> In cmake it can be set as `acada_ROOT/include/templates` (or something similar).
> But it's weird nonetheless.
> Anyhow, when I want to inform my users about some path, I would pass it through a cmake variable, or section in a `.pc` file.

Me too, I really would like to fix this as a CMake step during `find_package`, I`ll give it another try.

> You're right about `acadoCOpyTemplateFile`.
> The only other user `cpp` user of `TEMPLATE_PATHS` is this one: https://github.com/acado/acado/blob/b4e28f3131f79cadfd1a001e9fff061f361d3a0f/acado/code_generation/export_templated_file.cpp#L69
> Is that method called somewhere in the library?

Not in the lib, it is only used by the consumer if he wants to compile a codegenerator out of the provided include files.

**Comment by madebr** [comment]
> Not in the lib, it is only used by the consumer if he wants to compile a codegenerator out of the provided include files.

The `acado/code_generation` directory is also added to the library with a glob.
So why isn't it added to the library? How is acado supposed to be used maybe? Is there a runtime component?
https://github.com/acado/acado/blob/b4e28f3131f79cadfd1a001e9fff061f361d3a0f/acado/CMakeLists.txt#L93

**Comment by blackliner** [comment]
> > Not in the lib, it is only used by the consumer if he wants to compile a codegenerator out of the provided include files.
> 
> The `acado/code_generation` directory is also added to the library with a glob.
> So why isn't it added to the library? How is acado supposed to be used maybe? Is there a runtime component?
> https://github.com/acado/acado/blob/b4e28f3131f79cadfd1a001e9fff061f361d3a0f/acado/CMakeLists.txt#L93

I think you are right after all. Fixing the headerfile newer was enough... ACADO needs to be rebuilt EVERYTIME, the binaries cannot be distributed and are useless... damn

**Comment by madebr** [comment]
Just FYI, swig (=code generator for creating python/c#/java/... modules) also does some magic to get the path to its templates:
https://github.com/swig/swig/blob/b6c2438d7d7aac5711376a106a156200b7ff1056/Source/Modules/main.cxx#L935-L953
We also carry a patch to make it work on Linux: https://github.com/conan-io/conan-center-index/blob/master/recipes/swig/all/patches/0001-swig-linux-library-path.patch

In a nutshell, by default it assumes the templates to be in a fixed path relative to the executable.
Unless the path is defined explicitly by an environment variable.
